src/modules/sdbus/test/message.c

```diff
@@ -0,0 +1,472 @@
+        "successfully read back second (array) element");
+
+    ok (sd_bus_message_read (m, "b", &b) > 0
+        && b == 1,
+        "successfully read back third (boolean) element");
+
+    if (sd_bus_message_exit_container (m) <= 0)
+        BAIL_OUT ("error exiting struct container");
+
+    sd_bus_message_unref (m);
+    json_decref (o);
+}
+
+void test_variant (sd_bus *bus)
+{
+    json_t *o;
+    json_t *o2;
+    sd_bus_message *m;
+    int rc;
+
+    if (!(o = json_pack ("[[si][ss][sf]]",
+                         "i", 42,
+                         "s", "fubar",
+                         "d", -1.5)))
+        BAIL_OUT ("could not pack json array");
+    diagjson (o);
+
+    if (sd_bus_message_new (bus, &m, SD_BUS_MESSAGE_METHOD_CALL) < 0)
+        BAIL_OUT ("could not create method call message");
+
+    const char *fmt = "vvv";
+    rc = sdmsg_write (m, fmt, o);
+    ok (rc == 0,
+        "sdmsg_write works with variants");
+
+    if (sd_bus_message_seal (m, 42, 0) < 0
+        || sd_bus_message_rewind (m, true) < 0)
+        BAIL_OUT ("could not finalize message");
+
+    msgtype_is (m, "vvv");
+
+    diagmsg (m);
+
+    if (!(o2 = json_array ()))
+        BAIL_OUT ("could not create json array");
+    rc = sdmsg_read (m, fmt, o2);
+    diag ("sdmsg_read returned %d", rc);
+    ok (rc == 1,
+        "sdmsg_read works");
+    diagjson (o2);
+    ok (sd_bus_message_at_end (m, true),
+        "all message contents were read");
+    ok (json_equal (o, o2),
+        "json in/out are the same");
+
+    json_decref (o2);
+    sd_bus_message_unref (m);
+    json_decref (o);
+}
+
+void test_variant_as (sd_bus *bus)
+{
+    sd_bus_message *m;
+    json_t *in;
+    json_t *out;
+    const char *fmt = "v";
+    int rc;
+
+    if (!(in = json_pack ("[[s[sss]]]", "as", "foo", "bar", "baz")))
+        BAIL_OUT ("could not create json object");
+    if (sd_bus_message_new (bus, &m, SD_BUS_MESSAGE_METHOD_RETURN) < 0)
+        BAIL_OUT ("could not create message");
+    ok (sdmsg_write (m, fmt, in) == 0,
+        "sdmsg_write of variant string array works");
+
+    if (sd_bus_message_seal (m, 42, 0) < 0
+        || sd_bus_message_rewind (m, true) < 0)
+        BAIL_OUT ("could not finalize message");
+    diagmsg (m);
+    msgtype_is (m, fmt);
+
+    if (!(out = json_array ()))
+        BAIL_OUT ("could not create json array");
+
+    rc = sdmsg_read (m, fmt, out);
+    diag ("sdmsg_read returned %d", rc);
+    ok (rc == 1,
+        "sdmsg_read works on message containing string array variant");
+
+    diagjson (out);
+
+    const char *type;
+    json_t *val;
+    ok (json_unpack (out, "[[so]]", &type, &val) == 0
+        && streq (type, "as")
+        && json_is_array (val)
+        && json_array_size (val) == 3
+        && json_is_string (json_array_get (val, 0))
+        && json_is_string (json_array_get (val, 1))
+        && json_is_string (json_array_get (val, 2)),
+        "string array variant was translated properly");
```

Comment:
the anal part of me says to check the values too :-)

Reply:
This test now calls `json_equal()` on the input and output objects of json->dbus->json so the values are now checked.